### PR TITLE
Implement reading and writing boolean values

### PR DESCRIPTION
**Pull request by CoolCat467**
I know nothing in preforming server pings uses boolean values, but this would be super handy to have for handling responses with the `forgeData` section

### CODE REVIEW

**Comment by ItsDrike** [comment]
`forgeData` isn't a part of minecraft's official protocol, hence mcstatus doesn't support it. However `forgeData` seems to just be a part of the status response, as just an additional key. This means that it's already being obtained and is accessible, why would having a bool in the reader/writer classes help?

Mcstatus probably isn't the place to introduce these kinds of changes, as you're not expected to use these classes in your code. These classes are considered internal, and they are not a part of the public API. This means they're only expected to be used for internal functionalities, and if they're not necessary for those, they're not needed.

If you do need this kind of change, either implement it locally within your project, or consider checking out https://github.com/py-mine/mcproto, which already supports this behavior, and where the reader/writer classes are indeed a part of the public API: see ([this](https://github.com/py-mine/mcproto/blob/f2874148c18102efcf8746014749e77e0056b4ca/mcproto/protocol/base_io.py#L38), and [this](https://github.com/py-mine/mcproto/blob/f2874148c18102efcf8746014749e77e0056b4ca/mcproto/protocol/base_io.py#L104)).

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [inline mcstatus/protocol/connection.py:315]: `struct.unpack(">?", struct.pack(">?", True))` gives `(True,)`, so why not compare it to `True`?

```suggestion
        return self._unpack("?", self.read(1)) == True
```

The same for the async version.

---

This works now because `True == 1` (`bool` is a subclass of `int`).

**Review by PerchunPak — CHANGES_REQUESTED** [review]
I'm looking forward to the `forgeData` PR!

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [reply mcstatus/protocol/connection.py:315]: Why compare at all? Why not just `self._unpack("?", self.read(1))[0]`?

Comment (CoolCat467) [reply mcstatus/protocol/connection.py:315]: Basically because funny things with type annotations. We've annotated _unpack saying that it always returns an integer, even though in reality when we use "?" it actually returns a boolean value. The real question is should we fix _unpack to use overrides, or should we just pretend everything is an integer and just do kind of silly things like this to make mypy happy

**Comment by PerchunPak** [comment]
I agree with ItsDrike: before merging this PR, you need to explain why we even need this as `forgeData` is just a JSON attribute which we should parse without touching the `Connection` classes.

From [this comment](https://github.com/py-mine/mcstatus/issues/556#issuecomment-1566416444), I can see how you use the `Connection` classes to handle `forgeData` ([permalink to the code](https://github.com/CoolCat467/StatusBot/blob/34ffced1238f6a8ac754a1d1641f1544b90b5c90/src/StatusBot/decode_mods.py)). However, you don't use it to handle `forgeData`! It's used for the mysterious attribute `d` in the response body ([code](https://github.com/CoolCat467/StatusBot/blob/34ffced1238f6a8ac754a1d1641f1544b90b5c90/src/StatusBot/decode_mods.py#L83-L86))! This is undocumented in [the protocol docs](https://wiki.vg/Minecraft_Forge_Handshake) and with this name it's hard to find any information about what is it. I also thought that you modify the dict before passing to this function, but no. You pass it right after getting the status object.

We must be confident that we will even use the code in this PR, so you need to explain how will you use it. Currently, could you explain please what is `d` and what value it can have?

**Review by kevinkjt2000 — CHANGES_REQUESTED** [review]
Test changes are a requirement for new features. I concur with the other reviews too. We need to see examples of how this internal function would be used and I suggest that tests are a good way of doing that.

**Comment by ItsDrike** [comment]
> I concur with the other reviews too. We need to see examples of how this internal function would be used
@kevinkjt2000 

CoolCat467 already sent a link to an example of where this is useful in #556. However our internal API doesn't need it, and we don't expect others to use the `Connection` class, as it's not a part of the public API. So in my opinion, this only makes sense if we agree to #556, otherwise this PR can be closed.

(Here's the link to that example from the issue: https://github.com/CoolCat467/StatusBot/blob/main/src/StatusBot/decode_mods.py, it seems that forgeData dict holds some data under "d" key, which is a string in JSON, however it essentially just represents binary data, and the buffer/connection reader/writer classes are indeed necessary to read it).

I wouldn't mind adding this, but again, only if we agree to also add a parser for the forgeData, since otherwise this doesn't really serve us any purpose considering it wouldn't be used internally, and we don't expect it to be used externally.

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [reply mcstatus/protocol/connection.py:315]: Yeah, I like the cast more in this case.

Ultimately, overloads would be the best solution, but we'll probably migrate this library to eventually use `mcproto` for this low level networking, which already has this implemented properly (and indeed, that's done with overloads).

**Comment by CoolCat467** [comment]
> We must be confident that we will even use the code in this PR, so you need to explain how you will use it. Currently, could you explain please what is `d` and what value it can have?

Basically, the `d` field in the `forgeData` part of a ping response from a forge server is a UTF-16 string that represents binary data containing data like the forge mod loader network version, a big list of channels that all the forge mods use, and a list of mods the server has. Apparently forge ran into an issue a while ago with incredibly large modpacks sending a ton of data in those fields, so much that it exceeded the maximum message size the client can handle. To fix this, the `d` field was added, which basically acts like compressing the data so it doesn't exceed the max message size, and even if it tries to exceed the max message size, the server checks that and truncates the response if it would be too big.

More information is available in the docstrings in this file from forge itself: https://github.com/MinecraftForge/MinecraftForge/blob/42115d37d6a46856e3dc914b54a1ce6d33b9872a/src/main/java/net/minecraftforge/network/ServerStatusPing.java

Basically I need the ability to read boolean values because the compressed response contains boolean values. Then we could make a forgeData parser so to speak, but it wouldn't really be parsing just decoding the compressed response if the server has it. For older versions of forge, it just put all the data in the `forgeData` field and you didn't have to decompress it, so this wouldn't be needed for older versions.

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [inline mcstatus/protocol/connection.py:249]: You don't need to cast anything here.

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:713]: Why make this change here? This PR doesn't seem to be the right place for it, but also, it's not very necessary to write out all of these variables when they aren't being used. Might even cause some linters/language servers to report warnings due to unused variables.

File: mcstatus/protocol/connection.py
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:342]: This should be reverted, the type cast is not necessary in here, as `struct.unpack` returns `Any`, so there shouldn't be any issues with the return type-hint being `int`. This can be left as it was.

Is mypy reporting an issue here? We're using pyright in the codebase, which seems to be fine with this, which is why I'm confused as to why you even made this change here.

File: tests/protocol/test_connection.py
Comment (PerchunPak) [inline tests/protocol/test_connection.py:212]: We don't really need such basic tests, but if Kevin wants:

```suggestion
    @pytest.mark.parametrize("as_bytes,as_bool", [("01", True), ("00", False)])
    def test_read_bool(self, as_bytes: str, as_bool: bool) -> None:
        self.connection.receive(bytearray.fromhex(as_bytes))

        assert self.connection.read_bool() is as_bool

    @pytest.mark.parametrize("as_bytes,as_bool", [("01", True), ("00", False)])
    def test_write_bool(self, as_bytes: str, as_bool: bool) -> None:
        self.connection.write_bool(as_bool)

        assert self.connection.flush() == bytearray.fromhex(as_bytes)
```

P.S. I am unsure if pyright will be okay with `is as_bool`, you should check it.

**Review by ItsDrike — CHANGES_REQUESTED** [review]

**Review by PerchunPak — CHANGES_REQUESTED** [review]
Can we use `squash` strategy for this PR? As the commit history is not really clear.

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [reply mcstatus/protocol/connection.py:249]: Duplicate to https://github.com/py-mine/mcstatus/pull/555#discussion_r1209516063.

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [reply mcstatus/protocol/connection.py:540]: Duplicate to https://github.com/py-mine/mcstatus/pull/555#discussion_r1209515526, although we have different details, so I will not resolve this.

File: tests/protocol/test_connection.py
Comment (ItsDrike) [reply tests/protocol/test_connection.py:212]: Why wouldn't it be ok? You don't need to use literals when comparing using `is` , it's fine that `as_bool` is a variable.

Comment (PerchunPak) [reply tests/protocol/test_connection.py:212]: I remember that mypy/flake8/pylint/something doesn't allow using `is` with variables, because of the difference between statement and expression. I don't know how it works in pyright, so just mentioned this.

Comment (ItsDrike) [reply tests/protocol/test_connection.py:212]: I don't know about mypy or pylint, but flake8 (at least the default one, without considering extensions) doesn't have an issue with this, and pyright doesn't either ([quick example](https://pyright-playground.decorator-factory.su/?gzip=H4sIABH9dGQC_0srys9VSM7PyUlNLsnMzyvWS0xKVsjMLcgvKlFwTszJSUzKSeXiqrBSSMrPz1GwVXBLzClO5aqE80OKSlO50qzgaqOjY3XAcrFAyZzE3KSURCuoJq7E4uJUoLEVCpnFEH1QARAbJFaBoqISxkvT0MSUBQoCAGLoEbu7AAAA))

Comment (PerchunPak) [reply tests/protocol/test_connection.py:212]: Good, then just ignore that P.S.

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py:342]: Yes, I have mypy set to use `no-return-any`, I can revert this.

File: tests/protocol/test_connection.py
Comment (CoolCat467) [reply tests/protocol/test_connection.py:212]: Flake8 didn't like comparing a variable to literal boolean with `==`, it suggested changing to `is`

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py:713]: Basically because mypy was mad that the function was untyped

File: mcstatus/protocol/connection.py
Comment (CoolCat467) [reply mcstatus/protocol/connection.py]: Reverted in [64f7e0f](https://github.com/py-mine/mcstatus/pull/555/commits/64f7e0f49a5a95e1f346665510bdb18ca3f50d70)

Comment (PerchunPak) [reply mcstatus/protocol/connection.py]: You should use pyright as we use it for this project.

File: mcstatus/protocol/connection.py
Comment (PerchunPak) [reply mcstatus/protocol/connection.py]: (In case, if you are waiting for a review, friendly reminder that we still need to resolve this conversation)

Comment (CoolCat467) [reply mcstatus/protocol/connection.py]: Reverted changes as of 9676b71

**Review by ItsDrike — APPROVED** [review]
This can be merged, but only assuming we will later actually add support for showing `forgeData`. Currently, this will just add an unused method into Connection classes, and since they aren't officially a part of the public API, meaning we don't expect people to use these directly, it doesn't make sense to have it here without actually using it internally.

If this will come with another PR, adding public support for reading `forgeData`, this can be merged,

**Review by PerchunPak — APPROVED** [review]

**Comment by CoolCat467** [comment]
Yes, I plan to add a forgeData reader, but I would love to hear feedback about where the interface should be.

**Comment by PerchunPak** [comment]
> Yes, I plan to add a forgeData reader, but I would love to hear feedback about where the interface should be.

Inside [`JavaStatusResponse`](https://github.com/py-mine/mcstatus/blob/26b98c5421eead692996caf28953645e5a8b0f94/mcstatus/status_response.py#L102) class, of course.

**Comment by PerchunPak** [comment]
Or if you are about where to put the parsing logic, I can send you an example of how I imagine this architecture in Discord.

**Review by kevinkjt2000 — APPROVED** [review]

**Pull request merged**
